File: geeksgforgeeks/challenge0_fibonacci/fibonacci.java
/*
     Created by Avinash on 20th July, 2017
 */
import java.io.*;
import java.util.*;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;
public class fibonacci 
{

	static int memo[]; // instead of declaring the array size with some value, we can do it after taking the input, to save space
        public static void main(String[] args) 
	{
        	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int nth_term = 0,result = 0;
		try
		{       nth_term = Integer.parseInt(br.readLine());
			memo = new int[nth_term]; //array size declared here
			Arrays.fill(memo,-1);
		//	result = nth_fib_number_memoization(nth_term-1);
			result = nth_fib_number_tabulation(nth_term);
		
		}
		catch (IOException e)
		{
			e.printStackTrace();
		}
                System.out.println("The "+nth_term+"th term in Fibonacci series is "+result);
	}

        public static int nth_fib_number_memoization(int n)
	{
		if(memo[n] == -1)
         	{
         		if(n <= 1)
         			memo[n] = n;
         		else
         			memo[n] = nth_fib_number_memoization(n-1) + nth_fib_number_memoization(n-2);
         	}
         	return memo[n];	

        }
	public static int nth_fib_number_tabulation(int n)
	{
		int table[] = new int[n];
		table[0] = 0;
		table[1] = 1;
		for(int i = 2; i < n; i++)
		{
			table[i] = table[i-1] + table[i-2];
		}
		return table[n-1];
	}
}
         

File: geeksgforgeeks/challenge0_fibonacci/fibonacci.py
# Created by Avinash on 21st July, 2017
import sys
def nth_fibonacci_number_memoization(n, memo):
	if memo[n] == -1:
		if n <= 1:
			memo[n] = n
		else:
			memo[n] = nth_fibonacci_number_memoization(n-1,memo) + nth_fibonacci_number_memoization(n-2,memo);
	return memo[n];

def nth_fibonacci_number_tabulation(n, memo):
	memo[0] = 0
	memo[1] = 1
	for i in range(2, n+1):
			memo[i] = memo[i-1] + memo[i-2]
	return memo[n]

def nth_fibonacci_number_iterative(n):
	if n == 0:
		return 0
	term1 = 0
	term2 = 1
	temp = 0
	for i in range(2, n+1):
		temp = term2
		term2 = term1 + term2
		term1 = temp	
	return term2 

def process():
	n = int(input("Enter the fibonacci term "))
	memo = [-1]*(n+1)
	#print("The %dth Fibonacci term is %d" %(n, nth_fibonacci_number_memoization(n-1,memo)))
	#print("The %dth Fibonacci term is %d" %(n, nth_fibonacci_number_tabulation(n-1,memo)))
	print("The %dth Fibonacci term is %d" %(n, nth_fibonacci_number_iterative(n-1)))
	

if __name__=="__main__":
	process()